### PR TITLE
VM1: Create the Parser Module

### PR DESCRIPTION
**Pull request by wildmaples**
Additional information about the Parser Module can be found on the Nand2Tetris site: https://www.nand2tetris.org/project07

### CODE REVIEW

File: test/unit/parser_test.rb [hunk 1]
```diff
@@ -43,7 +43,7 @@ def test_inline_comments_in_input_file
     parser.advance
 
     # assert_equal("1234", parser.symbol)
-    # assert_equal(:A_COMMAND, parser.command_type)
+    assert_equal(:C_PUSH, parser.command_type)
```
Comment (tomstuart) [inline test/unit/parser_test.rb]: There’s another assertion in an earlier test (`# assert_equal(:C_COMMAND, parser.command_type)`) that should also be uncommented as part of this commit.

File: test/unit/parser_test.rb [hunk 2]
```diff
@@ -74,4 +74,28 @@ def test_command_type_returns_arithmetic_command
     parser.advance
     assert_equal(:C_ARITHMETIC, parser.command_type)
   end
+
```
Comment (tomstuart) [inline test/unit/parser_test.rb:77]: There are assertions in several earlier tests (e.g. `# assert_equal("push constant 17", parser.comp)`) which should probably be uncommented as part of this change.

File: test/unit/parser_test.rb [hunk 3]
```diff
@@ -98,4 +98,12 @@ def test_arg1_for_push_command
     parser.advance
     assert_equal("constant", parser.arg1)
   end
+
+  def test_arg2_for_push_command
+    input_file = StringIO.new("push constant 17")
+    parser = Parser.new(input_file)
+
+    parser.advance
+    assert_equal(17, parser.arg2)
```
Comment (tomstuart) [inline test/unit/parser_test.rb:107]: You could (if you want to) add `#arg2` assertions to some of the earlier tests that are incidentally checking the command contents as part of tests for `#has_more_commands?` etc.

**Review by tomstuart — APPROVED** [review]
Excellent! The parser works! There are still a few lines commented on in the tests, so I’ve left comments on the commits where those might be removed.

File: test/unit/parser_test.rb (hunk 3 above)
Comment (wildmaples) [reply test/unit/parser_test.rb:110]: Thankfully the same command was used... 😂 
```
    assert_equal(:C_PUSH, parser.command_type)
    assert_equal("constant", parser.arg1)
    assert_equal(17, parser.arg2)
```

**Pull request merged**